### PR TITLE
Add ebpf map cilium_egress_v4 for egress gateway

### PR DESCRIPTION
**Pull request by anfernee**
The map is of type lpm_trie, so that it supports range query for
destination CIDR. A packet that matches an entry in the map will
return an egress_info which contains the IP address of a gateway node
and an egress IP address. Gateway node IP represents the next hop for
this packet, while the egress IP represents the masqueraded source IP
address when the packet leaves the gateway node.

The commit contains a set of util subcommands for cilium, which provides
CRUD operations to the ebpf map:

```
cilium bpf egress {list,update,lookup,delete}
```

Signed-off-by: Yongkun Gui <ygui@google.com>

### CODE REVIEW

File: cilium/cmd/bpf_egress_delete.go
Comment (twpayne) [inline cilium/cmd/bpf_egress_delete.go]: This presumably should be `egressDeleteUsage`.

File: cilium/cmd/bpf_egress_lookup.go
Comment (twpayne) [inline cilium/cmd/bpf_egress_lookup.go]: This presumably should be `egressLookupUsage`.

File: cilium/cmd/bpf_egress_lookup.go
Comment (twpayne) [inline cilium/cmd/bpf_egress_lookup.go]: You can use `Args: cobra.ExactlyNArgs(2),` in the declaration of `bpfEgressLookupCmd` instead of manually checking `len(args)`.

This comment applies to other commands in this PR too.

File: cilium/cmd/bpf_egress_update.go
Comment (twpayne) [inline cilium/cmd/bpf_egress_update.go]: `egressUpdateUsage`.

File: bpf/lib/common.h
Comment (pchaigno) [inline bpf/lib/common.h:216]: Why is `egress_info` defined here and not in `bpf/lib/maps.h`?

File: cilium/cmd/bpf_egress.go
Comment (pchaigno) [inline cilium/cmd/bpf_egress.go:1]: We have some generated documentation for the CLI which will need to be updated. It's the reason why Travis CI is failing:
```
diff --git a/Documentation/cmdref/cilium_bpf.md b/Documentation/cmdref/cilium_bpf.md
index 8864e0aae..bc10ca6a4 100644
--- a/Documentation/cmdref/cilium_bpf.md
+++ b/Documentation/cmdref/cilium_bpf.md
@@ -23,6 +23,7 @@ Direct access to local BPF maps
 * [cilium](../cilium)	 - CLI
 * [cilium bpf bandwidth](../cilium_bpf_bandwidth)	 - BPF datapath bandwidth settings
 * [cilium bpf ct](../cilium_bpf_ct)	 - Connection tracking tables
+* [cilium bpf egress](../cilium_bpf_egress)	 - Manage the egress routing rules
 * [cilium bpf endpoint](../cilium_bpf_endpoint)	 - Local endpoint map
 * [cilium bpf fs](../cilium_bpf_fs)	 - BPF filesystem mount
 * [cilium bpf ipcache](../cilium_bpf_ipcache)	 - Manage the IPCache mappings for IP/CIDR <-> Identity
HINT: to fix this, run 'make -C Documentation update-cmdref'
```

File: cilium/cmd/bpf_egress_lookup.go
Comment (pchaigno) [inline cilium/cmd/bpf_egress_lookup.go]:
```suggestion
		common.RequireRootPrivilege("cilium bpf egress lookup <src_ip> <dest_ip>")
```

File: cilium/cmd/bpf_egress_list.go
Comment (pchaigno) [inline cilium/cmd/bpf_egress_list.go]: `kernelVersionWarning` has:
```
Note that for Linux kernel versions between 4.11 and 4.15 inclusive, the native
LPM map type used for implementing the IPCache does not provide the ability to
walk / dump the entries, so on these kernel versions this tool will never
return any entries, even if entries exist in the map. You may instead run:
    cilium map get cilium_ipcache
```
I think we need to adapt the message to refer to the egress map instead of the ipcache. Could be just a matter of passing it the map name:
```go
fmt.Sprintf(kernelVersionWarning, "egress map", egressmap.MapName)
```

File: cilium/cmd/bpf_egress_list.go
Comment (pchaigno) [inline cilium/cmd/bpf_egress_list.go]: I think we use uppercase for the column names in other CLI subcommands. We could also make these names a bit more informative, maybe `SRC IP & DST CIDR` and `GATEWAY INFO`?

File: pkg/maps/egressmap/egress.go
Comment (pchaigno) [inline pkg/maps/egressmap/egress.go]:
```suggestion
func (k *Key4) String() string            { return fmt.Sprintf("%s %s/%d", k.SourceIP, k.DestCIDR, k.PrefixLen) }
```

File: cilium/cmd/bpf_egress_lookup.go
Comment (pchaigno) [inline cilium/cmd/bpf_egress_lookup.go]: I would rename this subcommand to `get` to match the ipcache map. We can have `lookup` as an alias.

File: bpf/lib/maps.h
Comment (pchaigno) [inline bpf/lib/maps.h:172]: Could you add `egress_info` and `egress_key` to `bpf/bpf_alignchecker.c`? That will allow the Cilium agent to check on startup that there is no discrepancy between the user and kernel definitions of the map.

File: cilium/cmd/bpf_egress_update.go
Comment (pchaigno) [inline cilium/cmd/bpf_egress_update.go]: Maybe with `add` as an alias?

Comment (pchaigno) [inline cilium/cmd/bpf_egress_update.go]: Maybe with `add` as an alias?

File: pkg/maps/egressmap/egress.go
Comment (pchaigno) [inline pkg/maps/egressmap/egress.go:62]: I think there's something wrong here with the prefix length:
```
$ cilium bpf egress update 10.0.0.69 0.0.0.0/0 10.0.1.15 192.168.36.12
$ cilium bpf egress update 10.0.0.69 8.0.0.0/8 10.0.0.3 192.168.36.11
$ cilium bpf egress ls
key                    value
10.0.0.69 8.0.0.0 40   10.0.0.3 192.168.36.11    
10.0.0.69 0.0.0.0 32   10.0.1.15 192.168.36.12
```

File: cilium/cmd/bpf_egress_delete.go
Comment (pchaigno) [inline cilium/cmd/bpf_egress_delete.go:32]: It's sometimes useful to have a `flush` or `delete --all` command to flush the local state during incidents or when debugging. Can be done as a follow up though.

File: pkg/datapath/linux/config/config.go
Comment (pchaigno) [inline pkg/datapath/linux/config/config.go:148]: You will also need to update the documentation at https://docs.cilium.io/en/stable/concepts/ebpf/maps/#ebpf-maps.

**Review by pchaigno — CHANGES_REQUESTED** [review]
It looks good! Only minor comments below.

File: bpf/lib/common.h
Comment (anfernee) [reply bpf/lib/common.h:216]: I saw `endpoint_info` is listed here. I think it's because the pointer of this struct will be passed around as parameter.

File: cilium/cmd/bpf_egress_list.go
Comment (anfernee) [reply cilium/cmd/bpf_egress_list.go]: Good catch!

File: cilium/cmd/bpf_egress_lookup.go
Comment (anfernee) [reply cilium/cmd/bpf_egress_lookup.go]: +1

File: pkg/datapath/linux/config/config.go
Comment (anfernee) [reply pkg/datapath/linux/config/config.go:148]: Let's probably wait until the full feature merged. I'll add a TODO to remind myself.

File: pkg/maps/egressmap/egress.go
Comment (anfernee) [reply pkg/maps/egressmap/egress.go:62]: The prefix is actually 32 + the next cidr, because the first ip needs to be fully matched.

Comment (pchaigno) [reply pkg/maps/egressmap/egress.go:62]: Ah, right. Makes sense.

Could we remove 32 when we display then?

File: Documentation/cmdref/cilium-health_get.md
Comment (pchaigno) [inline Documentation/cmdref/cilium-health_get.md]: I think the `cilium-health` and `cilium-bugtool` changes were committed by mistake.

File: pkg/datapath/linux/config/config.go
Comment (pchaigno) [reply pkg/datapath/linux/config/config.go:148]: I think you said yesterday you were going to use https://github.com/cilium/cilium/issues/13575 to list PRs sent and PRs remaining to send. Maybe we could also track this TODO there, as a subitem. We tend to forget about TODOs added in the source code.

File: cilium/cmd/bpf_egress_delete.go
Comment (pchaigno) [reply cilium/cmd/bpf_egress_delete.go:32]: We could track this at https://github.com/cilium/cilium/issues/13575 as well, though probably not required for 1.10.

File: pkg/datapath/linux/config/config.go
Comment (anfernee) [reply pkg/datapath/linux/config/config.go:148]: Just did it :)

**Comment by anfernee** [comment]
For some reasons, the CI generates different doc from my machine.. 

```
diff --git a/Documentation/cmdref/cilium-health_status.md b/Documentation/cmdref/cilium-health_status.md
index 0b50d07..2261800 100644
--- a/Documentation/cmdref/cilium-health_status.md
+++ b/Documentation/cmdref/cilium-health_status.md
@@ -4,10 +4,6 @@
 
 Display cilium connectivity to other nodes
 
-### Synopsis
-
-Display cilium connectivity to other nodes
-
 cilium-health status [flags]
HINT: to fix this, run 'make -C Documentation update-cmdref'
```

**Comment by pchaigno** [comment]
> For some reasons, the CI generates different doc from my machine..

I'm guessing you don't have any local changes besides what you committed? Might be worth trying to delete the `cilium/docs-builder` container image so that it's rebuilt. Worst case, you can work around this by simply not committing the `Documentation/cmdref/cilium-{bugtool,health}*` changes.

**Comment by anfernee** [comment]
> > For some reasons, the CI generates different doc from my machine..
> 
> I'm guessing you don't have any local changes besides what you committed? Might be worth trying to delete the `cilium/docs-builder` container image so that it's rebuilt. Worst case, you can work around this by simply not committing the `Documentation/cmdref/cilium-{bugtool,health}*` changes.

Figured it out, I need to rebuild `cilium-health` and `cilium-bugtool` before running `make -C Documentation update-cmdref`.

**Review by pchaigno — APPROVED** [review]

**Comment by pchaigno** [comment]
test-me-please

File: cilium/cmd/bpf_egress_delete.go
Comment (jibi) [inline cilium/cmd/bpf_egress_delete.go]: Should we return an error in case `net.ParseIP()` fails to parse the source IP and returns `nil`?

File: cilium/cmd/bpf_egress_get.go
Comment (jibi) [inline cilium/cmd/bpf_egress_get.go]: Ditto here for checking if `args[0]` and `args[1]` are valid IPs

File: cilium/cmd/bpf_egress_update.go
Comment (jibi) [inline cilium/cmd/bpf_egress_update.go]: Same thing here for `net.ParseIP()`

File: pkg/maps/egressmap/egress.go
Comment (jibi) [inline pkg/maps/egressmap/egress.go:35]: Maybe add a comment in here stating that `PrefixLen` has to match the full `SourceIP` + the `DestCIDR`?

File: pkg/maps/egressmap/egress.go
Comment (jibi) [inline pkg/maps/egressmap/egress.go:47]: Perhaps I'm missing something, but why not just use a constant for that since this will always return 32 bits (alias `sizeof(ipv4) * 8`)?

**Review by jibi — COMMENTED** [review]
Just a few minor comments, but beside that looks good to me! :rocket:

File: pkg/maps/egressmap/egress.go
Comment (anfernee) [reply pkg/maps/egressmap/egress.go:47]: Actually I copied the same pattern from ipcache, which is pretty similar: 

https://github.com/cilium/cilium/blob/d9fa628eabe3568c2dee2405265e5ef2c2c7464d/pkg/maps/ipcache/ipcache.go#L60-L67

https://github.com/cilium/cilium/blob/d9fa628eabe3568c2dee2405265e5ef2c2c7464d/pkg/maps/ipcache/ipcache.go#L76-L81

**Comment by anfernee** [comment]
https://github.com/cilium/cilium/pull/14712/commits/09dfb115ca31c536d22f70a928c365958fccf223 is not related to this PR, but it silents the CI.

**Comment by anfernee** [comment]
test-me-please

**Review by jibi — APPROVED** [review]

**Comment by pchaigno** [comment]
> [09dfb11](https://github.com/cilium/cilium/commit/09dfb115ca31c536d22f70a928c365958fccf223) is not related to this PR, but it silents the CI.

We recently broke one of the GitHub Actions. https://github.com/cilium/cilium/pull/14778 will fix it; you can rebase on top of master once that PR is merged.

**Comment by anfernee** [comment]
> > [09dfb11](https://github.com/cilium/cilium/commit/09dfb115ca31c536d22f70a928c365958fccf223) is not related to this PR, but it silents the CI.
> 
> We recently broke one of the GitHub Actions. #14778 will fix it; you can rebase on top of master once that PR is merged.

I see. Done.

**Comment by pchaigno** [comment]
test-me-please

**Comment by anfernee** [comment]
The failure seems to be due to flaky test: ```
17:11:29 STEP: Making 10 HTTP requests from k8s3:60000 to "tftp://192.168.36.11:32153/hello"
17:11:31 STEP: Making 10 HTTP requests from k8s3:60000 to "tftp://192.168.36.12:32153/hello"
17:11:32 STEP: Making 10 HTTP requests from k8s3:61000 to "tftp://192.168.36.11:32153/hello"
17:11:35 STEP: Making 10 HTTP requests from k8s3:61000 to "tftp://192.168.36.12:32153/hello"
17:11:38 STEP: Making 10 HTTP requests from k8s3:62000 to "tftp://192.168.36.11:32153/hello"
17:11:41 STEP: Making 10 HTTP requests from k8s3:62000 to "tftp://192.168.36.12:32153/hello"
FAIL: Cannot connect to service "tftp://192.168.36.12:32153/hello" (5/10)
Expected command: kubectl exec -n kube-system log-gatherer-7g6rd -- curl --path-as-is -s -D /dev/stderr --fail --connect-timeout 5 --max-time 20 --local-port 62000 tftp://192.168.36.12:32153/hello -w "time-> DNS: '%{time_namelookup}(%{remote_ip})', Connect: '%{time_connect}',Transfer '%{time_starttransfer}', total '%{time_total}'" | grep 'Hostname:'  
To succeed, but it failed:
Exitcode: 1 
Err: exit status 1
Stdout:
 	 
Stderr:
 	 command terminated with exit code 72
```

**Comment by pchaigno** [comment]
test-only --focus="K8sServicesTest Checks service across nodes Tests NodePort BPF Tests with vxlan Tests NodePort with Maglev Tests Maglev backend selection" --kernel_version=net-next --k8s_version=1.13

**Comment by twpayne** [comment]
retest-net-next

**Comment by pchaigno** [comment]
> retest-net-next

@twpayne See the above discussion. The failure is a flake, confirmed by test-only command.

File: pkg/maps/egressmap/doc.go
Comment (twpayne) [inline pkg/maps/egressmap/doc.go]: nit: Copyright year should be 2021.

File: pkg/maps/egressmap/egress.go
Comment (twpayne) [inline pkg/maps/egressmap/egress.go]: nit: Copyright year should be 2021.

**Comment by pchaigno** [comment]
test-me-please

**Comment by anfernee** [comment]
These 2 tests seem to be flake:
```
Test Result (2 failures / +2)
Suite-runtime.RuntimePolicyImportTests checks policy trace output
Suite-runtime.RuntimeFQDNPolicies Enforces ToFQDNs policy
```

What command to run to retest them?

**Comment by pchaigno** [comment]
test-runtime

**Comment by anfernee** [comment]
looks like it all passed.

**Pull request merged**
